library/std/src/os/unix/net/addr.rs

```diff
@@ -127,6 +127,73 @@ impl SocketAddr {
         Ok(SocketAddr { addr, len })
     }
 
+    /// Constructs a `SockAddr` with the family `AF_UNIX` and the provided path.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the path is longer than `SUN_LEN`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(unix_socket_creation)]
+    /// use std::os::unix::net::SocketAddr;
+    /// use std::path::Path;
+    ///
+    /// # fn main() -> std::io::Result<()> {
+    /// let address = SocketAddr::unix("/path/to/socket")?;
+    /// assert_eq!(address.as_pathname(), Some(Path::new("/path/to/socket")));
+    /// # Ok(())
+    /// # }
```

Comment:
This could also use an example for abstract sockets to demonstrate we can create `Path` instances containing NUL, which might be surprising since they're not valid characters for paths.

Reply:
If we want to steer users towards a separate method for abstract paths instead then should this explicitly reject NUL bytes?

Reply:
> If we want to steer users towards a separate method for abstract paths instead then should this explicitly reject NUL bytes?

Yes. See discussion above. All existing functions taking a `Path`, including the already existing (but private) `Path` to sockaddr function, return an error on embedded nul bytes.